cranelift/interpreter/src/interpreter.rs

```diff
@@ -257,38 +275,93 @@ impl<'a> State<'a, DataValue> for InterpreterState<'a> {
-            Ok(unsafe { DataValue::read_value_from(pointer, ty) })
-        } else {
-            Err(MemoryError::InsufficientMemory(offset, self.heap.len()))
+    fn stack_address(
+        &self,
+        size: AddressSize,
+        slot: StackSlot,
+        offset: u64,
+    ) -> Result<Address, MemoryError> {
+        let stack_slots = &self.get_current_function().stack_slots;
+        let stack_slot = &stack_slots[slot];
+
+        // offset must be `0 <= Offset < sizeof(SS)`
+        if offset >= stack_slot.size as u64 {
+            return Err(MemoryError::InvalidOffset {
+                offset,
+                max: stack_slot.size as u64,
+            });
         }
-    }
 
-    fn store_heap(&mut self, offset: usize, v: DataValue) -> Result<(), MemoryError> {
-        if offset + 16 < self.heap.len() {
-            let pointer = self.heap[offset..offset + 16].as_mut_ptr() as *mut _ as *mut u128;
-            Ok(unsafe { v.write_value_to(pointer) })
-        } else {
-            Err(MemoryError::InsufficientMemory(offset, self.heap.len()))
-        }
+        // Calculate offset from the base of the stack, to the current frame
+        let frame_offset: u64 = self
+            .frame_stack
+            .iter()
+            .enumerate()
+            // Remove the last frame, which is the frame that we are currently executing
+            .filter(|&(i, _)| i != self.frame_stack.len() - 1)
+            // Sum all of the stack slots from the previous frames
+            .flat_map(|(_, frame)| frame.function.stack_slots.values())
+            .map(|v| v.size as u64)
+            .sum();
+
+        // Calculate the offset from the current frame to the requested stack slot
+        let slot_offset: u64 = stack_slots
+            .keys()
+            .filter(|k| k < &slot)
+            .map(|k| stack_slots[k].size as u64)
+            .sum();
+
+        let final_offset = frame_offset + slot_offset + offset;
+        Address::from_parts(size, AddressRegion::Stack, 0, final_offset)
+    }
+
+    fn heap_address(&self, _size: AddressSize, _offset: u64) -> Result<Address, MemoryError> {
+        unimplemented!()
     }
 
-    fn load_stack(&self, _offset: usize, _ty: Type) -> Result<DataValue, MemoryError> {
-        unimplemented!()
+    fn checked_load(&self, addr: Address, ty: Type) -> Result<DataValue, MemoryError> {
+        let load_size = ty.bytes() as usize;
+
+        let raw_addr = match addr.region {
+            AddressRegion::Stack => {
+                let addr_start = addr.offset as usize;
+                let addr_end = addr_start + load_size;
+                if addr_end > self.stack.len() {
+                    return Err(MemoryError::OutOfBoundsLoad { addr, load_size });
+                }
+
+                self.stack[addr_start..addr_end].as_ptr() as *const _ as *const u128
+            }
+            _ => unimplemented!(),
+        };
+
+        Ok(unsafe { DataValue::read_value_from(raw_addr, ty) })
```

Comment:
Converting to a raw pointer then accessing it with `unsafe` feels a little bit unnecessary to me -- not that it's hard to see the safety here, since the code is very simple, but it entrenches a pattern that could go wrong later if it gets much more complicated.

Instead perhaps could we:
- Get a slice `&mut [u8]` of raw memory bytes from whatever region/object we're accessing (so the underlying interpreter-state for the stack, any wasm memories, etc., is just a `Vec<u8>`)
- Slice that into the number of bytes we need at the offset we need
- Provide accessors on `DataValue` that read and write from `&[u8]` / `&mut [u8]`?

Reply:
I think @afonso360 was just going along with what existed previously in `DataValue` (those unsafe methods are necessary for interacting with trampolines). I agree that it would probably be better to just some safe versions of those functions to `DataValue`.

Reply:
Indeed, it was a reasonable approach given the APIs that existed! The unsafe accessors on `DataValue` can certainly stay for the other use-cases (and perhaps a future refactor can centralize those into e.g. `unsafe { std::slice::from_raw_parts(...) }` in the trampoline-adjacent code and then use of safe APIs on `DataValue`, but that's a separate conversation!).

Reply:
It looks like `read_value_from`/`write_value_to` are only used in the interpreter, so I replaced them with `write_to_slice` and `read_from_slice`. It looks like we don't use them anywhere else, but if we need them we have a version that uses `slice::from_raw_parts` in 5c594c4

Reply:
Oh, it looks like we duplicate the `write_value_to` in function_runner.rs

https://github.com/bytecodealliance/wasmtime/blob/b2bcdd13ecb87e9c7cdae9cead1e490bcd47e4fe/cranelift/filetests/src/function_runner.rs#L235-L247

Reply:
Since we had this duplicated, I thought it would be best to merge it all into DataValue right now before we have some hard to track down memory layout mismatch.